### PR TITLE
DFR-2980-Mandatory-Postcode-Field

### PR DESCRIPTION
**Pull request by dawudgovuk**
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/DFR-2980

### Change description ###
CT - Postcode field not mandatory - Update contact details

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 1]
```diff
@@ -62,7 +68,119 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+        data.getContactDetailsWrapper().setRespondentSolicitorAddress(addressBlank);
+
+        handler.canHandle(MID_EVENT, CONTESTED, UPDATE_CONTACT_DETAILS);
+        GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle = handler.handle(finremCallbackRequest, AUTH_TOKEN);
+
+        assertEquals(1, handle.getErrors().size());
+        assertTrue(handle.getErrors().contains("Postcode field is required for respondent solicitor address."));
+
+    }
 }
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:186]: I’d prefer to add a new line at the end of this file.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 2]
```diff
@@ -62,7 +68,119 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+        FinremCaseData data = caseDetails.getData();
+
+        Address address = new Address();
+        address.setPostCode("AB1 1AB");
+        data.getContactDetailsWrapper().setApplicantAddress(address);
+        data.getContactDetailsWrapper().setRespondentAddress(address);
+        data.getContactDetailsWrapper().setApplicantRepresented(YesOrNo.NO);
+        data.getContactDetailsWrapper().setContestedRespondentRepresented(YesOrNo.NO);
+
+        handler.canHandle(MID_EVENT, CONTESTED, UPDATE_CONTACT_DETAILS);
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:89]: I don't think this line of code is needed.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 3]
```diff
@@ -62,7 +68,119 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+        Address address = new Address();
+        address.setPostCode("AB1 1AB");
+        Address addressBlank = new Address();
+        addressBlank.setPostCode("");
+        data.getContactDetailsWrapper().setApplicantAddress(address);
+        data.getContactDetailsWrapper().setRespondentAddress(addressBlank);
+        data.getContactDetailsWrapper().setApplicantRepresented(YesOrNo.NO);
+        data.getContactDetailsWrapper().setContestedRespondentRepresented(YesOrNo.NO);
+
+        handler.canHandle(MID_EVENT, CONTESTED, UPDATE_CONTACT_DETAILS);
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:131]: same as above comment.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 4]
```diff
@@ -62,7 +68,119 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+        address.setPostCode("AB1 1AB");
+        Address addressBlank = new Address();
+        addressBlank.setPostCode("");
+        data.getContactDetailsWrapper().setApplicantAddress(address);
+        data.getContactDetailsWrapper().setRespondentAddress(address);
+        data.getContactDetailsWrapper().setApplicantRepresented(YesOrNo.YES);
+        data.getContactDetailsWrapper().setContestedRespondentRepresented(YesOrNo.YES);
+        data.getContactDetailsWrapper().setRespondentSolicitorAddress(addressBlank);
+
+        handler.canHandle(MID_EVENT, CONTESTED, UPDATE_CONTACT_DETAILS);
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:179]: same as above comment.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 5]
```diff
@@ -62,7 +68,119 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+    public void givenConsentedCase_WhenEmptyApplicantSolicitorPostCode_thenHandlerWillShowMessage() {
+
+        FinremCallbackRequest finremCallbackRequest = buildCallbackRequest();
+        FinremCaseDetails caseDetails = finremCallbackRequest.getCaseDetails();
+        FinremCaseData data = caseDetails.getData();
+
+        Address address = new Address();
+        address.setPostCode("AB1 1AB");
+        Address addressBlank = new Address();
+        addressBlank.setPostCode("");
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:149]: should we add an extra test case for null postCode?

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandler.java [hunk 6]
```diff
@@ -38,7 +48,46 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(Finrem
...
+    }
+
+    private List<String> validateCaseData(FinremCaseData caseData) {
+
+        List<String> errors = new ArrayList<>();
+        ContactDetailsWrapper wrapper = caseData.getContactDetailsWrapper();
+
+        if (caseData.isApplicantRepresentedByASolicitor()) {
+            if (wrapper.getApplicantSolicitorAddress() != null
+                && ObjectUtils.isEmpty(wrapper.getApplicantSolicitorAddress().getPostCode()))  {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandler.java:65]: It's not a big problem but it can be made it more readable.

such as 
`Optional.ofNullable(wrapper.getApplicantSolicitorAddress())
            .map(Address::getPostCode)
            .map(String::isEmpty)
            .orElse(false)`

https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=1756&id=finrem-case-orchestration&open=AZB4fiLN2_G31vFDdufu

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandler.java [hunk 7]
```diff
@@ -38,7 +48,46 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(Finrem
...
+            .data(caseData).errors(errors).build();
+    }
+
+    private List<String> validateCaseData(FinremCaseData caseData) {
+
+        List<String> errors = new ArrayList<>();
+        ContactDetailsWrapper wrapper = caseData.getContactDetailsWrapper();
+
+        if (caseData.isApplicantRepresentedByASolicitor()) {
+            if (wrapper.getApplicantSolicitorAddress() != null
```
Comment (hamzahtahirhmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandler.java:64]: Just a thought on this solution. 

If the solicitor's postcode is provided, the backend will continue to the next check, which is the applicant's address. If the applicant's postcode is missing, an error may be raised prematurely, even though we haven't navigated to the applicant details page yet due to checking each field sequentially. Is this correct?

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 8]
```diff
@@ -62,7 +68,194 @@ private FinremCallbackRequest buildCallbackRequest() {
             .builder()
             .eventType(EventType.UPDATE_CONTACT_DETAILS)
             .caseDetails(FinremCaseDetails.builder().id(123L).caseType(CONTESTED)
-                .data(FinremCaseData.builder().ccdCaseType(CONTESTED).build()).build())
+                .data(FinremCaseData.builder().ccdCaseType(CONTESTED).contactDetailsWrapper(new ContactDetailsWrapper()).build()).build())
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:71]: The `.contactDetailsWrapper(new ContactDetailsWrapper())` isn't required

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 9]
```diff
@@ -62,7 +68,194 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+        address.setPostCode("AB1 1AB");
+        data.getContactDetailsWrapper().setApplicantAddress(address);
+        data.getContactDetailsWrapper().setRespondentAddress(address);
+        data.getContactDetailsWrapper().setApplicantRepresented(YesOrNo.NO);
+        data.getContactDetailsWrapper().setContestedRespondentRepresented(YesOrNo.NO);
+
+        GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle = handler.handle(finremCallbackRequest, AUTH_TOKEN);
+        assertEquals(0, handle.getErrors().size());
+    }
+
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:92]: Extra new line not required

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 10]
```diff
@@ -62,7 +68,194 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+
+    @Test
+    public void givenConsentedCase_WhenEmptyApplicantPostCode_thenHandlerWillShowMessage() {
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:95]: No need for public

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java [hunk 11]
```diff
@@ -62,7 +68,194 @@ private FinremCallbackRequest buildCallbackRequest() {
...
+
+        Address address = new Address();
+        address.setPostCode("");
+        data.getContactDetailsWrapper().setApplicantAddress(address);
+        data.getContactDetailsWrapper().setApplicantRepresented(YesOrNo.NO);
+
+        GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle = handler.handle(finremCallbackRequest, AUTH_TOKEN);
+
+        assertEquals(1, handle.getErrors().size());
+        assertTrue(handle.getErrors().contains("Postcode field is required for applicant address."));
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UpdateContactDetailsContestedMidHandlerTest.java:109]: You should make your unit tests as precise as possible. You can use 
handle.getErrors.get(0) rather than use contains (other instances of this in the other tests)

**Review by jthmcts — APPROVED** [review]

**Review by hamzahtahirhmcts — APPROVED** [review]
